### PR TITLE
polyfempy

### PR DESCRIPTION
**Pull request by teseoch**
c/c++ `@conda-forge/help-c-cpp`

### CODE REVIEW

**Comment by SylvainCorlay** [comment]
Please create a new folder including your recipe instead of modifying the example recipe in-place!

**Comment by teseoch** [comment]
It is a new folder but I used `git mv. I will fix that

**Comment by teseoch** [comment]
@beckermr I fixed the folder problem.
Btw, do you have any idea of why `staged-recipes (win_64 win)` is not passing?

Last week it was and I havent changed anything in the windows build.

Thank you

**Comment by teseoch** [comment]
for the life of me I cannot understand how to fix the build on linux.
The problem is that my lib depends on openmp and when the package is imported in test it cannot find a symbol. Any idea?

File: recipes/polyfempy/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    - "{{ PYTHON }} -m pip install . --no-deps -vv"
+
+requirements:
+  build:
+    # if your project compiles code (such as a C extension) then add the required compilers as separate entries here.
+    # compilers are named 'c', 'cxx' and 'fortran'.
+    - {{ compiler('cxx') }}
+    - git
+    - cmake
+    - openmp  # [linux]
```
Comment (beckermr) [inline recipes/polyfempy/meta.yaml]: you need llvm-openmp here. I know this is confusing. Sorry!

File: recipes/polyfempy/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  host:
+    - python
+    - pip
+    - setuptools
+    - openmp  # [linux]
```
Comment (beckermr) [inline recipes/polyfempy/meta.yaml]: Ditto here.

File: recipes/polyfempy/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  run:
+    - python
+    - numpy
```
Comment (beckermr) [inline recipes/polyfempy/meta.yaml]: This should be a pin_compatible statement.

File: recipes/polyfempy/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    - openmp  # [linux]
```
Comment (beckermr) [inline recipes/polyfempy/meta.yaml]: llvm-openmp here

File: recipes/polyfempy/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  # Once resolved, it should be removed.
+  skip: True  # [py<31]
+  script:
+    - "{{ PYTHON }} -m pip install . --no-deps -vv"
+
+requirements:
+  build:
+    # if your project compiles code (such as a C extension) then add the required compilers as separate entries here.
+    # compilers are named 'c', 'cxx' and 'fortran'.
+    - {{ compiler('cxx') }}
```
Comment (beckermr) [inline recipes/polyfempy/meta.yaml]: You may need add the c compiler too? IDK. Once you make the changes below, def worth retrying.

File: recipes/polyfempy/meta.yaml (hunk 1 above)
Comment (teseoch) [reply recipes/polyfempy/meta.yaml]: ok thanks. I saw comments about llvm but I was convinced that I am using gcc so no need, sorry

File: recipes/polyfempy/meta.yaml (hunk 3 above)
Comment (teseoch) [reply recipes/polyfempy/meta.yaml]: what is a pin_compatible statement?

File: recipes/polyfempy/meta.yaml [hunk 6]
```diff
@@ -14,69 +9,49 @@ package:
...
 
 requirements:
   build:
-    # if your project compiles code (such as a C extension) then add the required compilers as separate entries here.
-    # compilers are named 'c', 'cxx' and 'fortran'.
     - {{ compiler('cxx') }}
     - git
     - cmake
+    - llvm-openmp  # [linux]
     - openmp  # [linux]
```
Comment (beckermr) [inline recipes/polyfempy/meta.yaml]: Remove the openmp package too. The openmp package is not for GCC.

File: recipes/polyfempy/meta.yaml (hunk 3 above)
Comment (beckermr) [reply recipes/polyfempy/meta.yaml]: See the docs here: https://conda-forge.org/docs/meta.html#building-against-numpy

It helps keep everything pinned correctly.

**Comment by beckermr** [comment]
Ack. I forgot to say that you only need to specify llvm-openmp on osx! Sorry!

**Comment by teseoch** [comment]
> Ack. I forgot to say that you only need to specify llvm-openmp on osx! Sorry!

That was my understanding. On unix it uses gcc which has openmp already. On osx there 2 version of llvm with/without openmp.

I will remove all openmp dependencies since on osx is not necessary and on unix openmp is the python version and not the c version

File: recipes/polyfempy/meta.yaml (hunk 5 above)
Comment (teseoch) [reply recipes/polyfempy/meta.yaml]: tried that, now working. Any other idea?

**Comment by beckermr** [comment]
Hmmmm....

```
Cloning into 'polyfem'...
  HEAD is now at e4c857d hacky geogram static lib
  -- GCC >= 4.9 detected, enabling colored diagnostics
  Cloning into 'spdlog'...
  HEAD is now at 188cff7 Merge pull request #831 from taketwo/fix-830
  -- Build type: Release
  Cloning into 'tinyfiledialogs'...
  HEAD is now at 511e650 v3.3.7
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
  ModuleNotFoundError: No module named 'sympy'
  CMake Warning at 3rdparty/polyfem/cmake/PolyfemUtils.cmake:53 (MESSAGE):
    Unable to find sympy and/or numpy, eigs.py not running
  Call Stack (most recent call first):
    3rdparty/polyfem/CMakeLists.txt:143 (polyfem_autogen)
```

So a few things:

1. It appears to be missing sympy. Maybe add this to the host section?
2. cmake looks like it is downloading other dependencies? Generally bundling/vendoring software like this on conda-forge is not a great idea. Instead it is better to make separate packages.

**Comment by teseoch** [comment]
1. Scipy and numpy are optional. That project generates some file with python but the generated source is also in the repo so it is not necessary 

2. That project has many depedencies which are not mine (and dont want to maintain) and are cpp only. I don’t think it is a good idea to add them to conda. The problem is that one of them (geogram) depends on openmp and that’s creating the *linker* problem.

**Comment by beckermr** [comment]
That all sounds good except that the error above is for importing `sympy` which is neither numpy nor scipy.

**Comment by teseoch** [comment]
Should b scipy, i did a typo

**Comment by beckermr** [comment]
Mmmk. Well this is all I got. Maybe someone else from @conda-forge/staged-recipes can help?

**Comment by scopatz** [comment]
According to [this](https://github.com/polyfem/polyfem/blob/master/cmake/PolyfemUtils.cmake) you need to add sympy (and maybe numpy) to the host requirements

**Comment by teseoch** [comment]
It is not that. I wrote that cmake. 

My code is using python to generate some cpp files. I made that optional to simplify to build (and because the python code didn't changed anymore) and added the generated code to the repository.

If this would be a problem, the cmake (or compilation) would fail. The code compiles and generates the binary. It is a **linker** problem, when the code is called it cannot find the dynamic library openmp.

The main library (PolyFEM) has CI that passes on the 3 os:
https://github.com/polyfem/polyfem/

File: recipes/polyfempy/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+source:
+  url: https://github.com/polyfem/polyfem-python/archive/{{ version }}.tar.gz
+  sha256: e24bff3fd6c4a54f58be5af3be9f3038f378fed5b250a1da95ab75543d549d5c
+
+build:
+  number: 0
+  skip: True  # [py<31]
+  script:
+    - "{{ PYTHON }} -m pip install . --no-deps -vv"
```
Comment (scopatz) [inline recipes/polyfempy/meta.yaml]: The repo doesn't seem to have a toml file or setup.py, so I am not sure what this is doing

**Comment by scopatz** [comment]
Yeah, I agree it is a linker problem. Your CMake configuration doesn't seem to set the linker flags to use OpenMP (that I could find) and you are not setting `LD_FLAGS` in the recipe because you are just using `pip install`.

File: recipes/polyfempy/meta.yaml (hunk 7 above)
Comment (teseoch) [reply recipes/polyfempy/meta.yaml]: The build is a little complicated.

I have 3 repos:
1) with the pure cpp library https://github.com/polyfem/polyfem/
2) with the binding using 1 and pybind11 which contains the setup https://github.com/polyfem/polyfem-python
3) the fork where i have the recipe

**Comment by teseoch** [comment]
Yes I agree. I am trying to cleanup my cmake for openmp.

The problem is that openmp is not a library like all others

**Comment by teseoch** [comment]
Yeah! Thanks all for the suggestion:)

@beckermr can you approve

**Pull request merged**